google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -381,32 +388,71 @@ private static boolean isWindows() {
   }
 
   protected static String getAppEngineProjectId() {
-    try {
-      Class<?> factoryClass =
-          Class.forName("com.google.appengine.api.appidentity.AppIdentityServiceFactory");
-      Class<?> serviceClass =
-          Class.forName("com.google.appengine.api.appidentity.AppIdentityService");
-      Method method = factoryClass.getMethod("getAppIdentityService");
-      Object appIdentityService = method.invoke(null);
-      method = serviceClass.getMethod("getServiceAccountName");
-      String serviceAccountName = (String) method.invoke(appIdentityService);
-      int indexOfAtSign = serviceAccountName.indexOf('@');
-      return serviceAccountName.substring(0, indexOfAtSign);
-    } catch (ClassNotFoundException exception) {
-      if (System.getProperty("com.google.appengine.runtime.version") != null) {
-        // Could not resolve appengine classes under GAE environment.
-        throw new RuntimeException("Google App Engine runtime detected "
-            + "(the environment variable \"com.google.appengine.runtime.version\" is set), "
-            + "but unable to resolve appengine-sdk classes. "
-            + "For more details see "
-            + "https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/APPENGINE.md");
+    if (PlatformInformation.isOnGAEStandard7()) {
+      try {
+        Class<?> factoryClass =
+            Class.forName("com.google.appengine.api.appidentity.AppIdentityServiceFactory");
+        Class<?> serviceClass =
+            Class.forName("com.google.appengine.api.appidentity.AppIdentityService");
+        Method method = factoryClass.getMethod("getAppIdentityService");
+        Object appIdentityService = method.invoke(null);
+        method = serviceClass.getMethod("getServiceAccountName");
+        String serviceAccountName = (String) method.invoke(appIdentityService);
```

Comment:
I decided to remove reflection on appengine APIs, because the system prop approach doesn't have "thread access" issue and should work on both dev and prod servers for GAE J7.